src/app/scanning.service.ts

```diff
@@ -0,0 +1,167 @@
+}
+
+export class ScanningModel {
+  currentHighlight: Scannable;
+  currentSelection: Scannable;
+}
+
+export interface ScannableCollectionProvider extends Subscriber<ScanningModel> {
+  getScannableCollections(): ScannableCollection[];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ScanningService {
+
+  private topLevelScannables: ScannableCollection;
+  private currentCollection: ScannableCollection;
+  private scanningModel: ScanningModel;
+  private currentSelectedIndex = 0;
+  private observers: ScannableCollectionProvider[];
+  private intervalId;
+
+  constructor(private configService: ConfigService) {
+    this.topLevelScannables = new ScannableCollection(0, '');
+    this.currentCollection = this.topLevelScannables;
+    this.scanningModel = new ScanningModel();
+    this.observers = [];
+  }
+
+  getScanningModel(): Observable<ScanningModel> {
+    return new Observable<ScanningModel>(this.registerScannable);
+  }
+
+  registerScannable = (observer: ScannableCollectionProvider): Subscription => {
+    this.observers.push(observer);
+
+    observer.getScannableCollections().forEach(col => col.sortChildren());
+
+    this.topLevelScannables.getChildren().push(...observer.getScannableCollections());
+    this.topLevelScannables.sortChildren();
+    this.startScanning();
+
+    return new Subscription(() => this._unsubscribe(observer));
+  }
+
+  _unsubscribe(observer: ScannableCollectionProvider) {
+    this.observers.splice(this.observers.indexOf(observer), 1);
+    observer.getScannableCollections().forEach(scannable => {
+      const collection = <ScannableCollection> scannable;
+      this.topLevelScannables.getChildren().splice(this.topLevelScannables.getChildren().indexOf(collection), 1);
+    });
+
+    this.currentCollection = this.topLevelScannables;
+
+    if (this.observers.length === 0) {
+      clearInterval(this.intervalId);
+      this.intervalId = undefined;
+      this.scanningModel = new ScanningModel();
+    }
+  }
+
+  startScanning() {
+    this.currentSelectedIndex = 0;
+
+    if (this.intervalId === undefined && this.observers.length >= 1 && this.configService.scanningConfig.enabled) {
+      this.intervalId = setInterval(this.updateHighlighted, this.configService.scanningConfig.time);
+    }
+  }
+
+  updateHighlighted = () => {
+    if (this.currentSelectedIndex >= this.currentCollection.getChildren().length) {
+      this.currentSelectedIndex = 0;
+    }
+
+    this.scanningModel.currentHighlight = this.currentCollection.getChild(this.currentSelectedIndex++);
+    this.observers.forEach(observer => observer.next(this.scanningModel));
+  }
+
+  updateSelected = () => {
+    const current = this.scanningModel.currentHighlight;
+    this.scanningModel.currentSelection = current;
+    this.scanningModel.currentHighlight = undefined;
+    this.observers.forEach(observer => observer.next(this.scanningModel));
+    clearInterval(this.intervalId);
+    this.intervalId = undefined;
+
+    setTimeout(() => {
```

Comment:
Would it be easier and safer to handle this with some state (i.e. a flag!) to just do something different in a single timer function?

Reply:
I guess we might have to wait two `ticks` to avoid clearing the selection if the timer ticks just straight after this event.

I'm concerned about the interval timer being stopped and started from more than one method